### PR TITLE
Add QgsTextRenderer method to render text along a line/curved path

### PR DESCRIPTION
**Pull request by nyalldawson**

### CODE REVIEW

File: src/core/textrenderer/qgstextrenderer.cpp [hunk 1]
```diff
@@ -150,6 +152,215 @@ void QgsTextRenderer::drawText( QPointF point, double rotation, Qgis::TextHorizo
...
+
+  QVector< double > characterWidths( graphemes.count() );
+  QVector< double > characterHeights( graphemes.count() );
+  QVector< double > characterDescents( graphemes.count() );
+  QFont previousNonSuperSubScriptFont;
+
+  for ( int i = 0; i < graphemes.count(); i++ )
+  {
+    // reconstruct how Qt creates word spacing, then adjust per individual stored character
+    // this will allow PAL to create each candidate width = character width + correct spacing
```
Comment (nirvn) [inline src/core/textrenderer/qgstextrenderer.cpp:216]: Is the PAL reference here correct? I assume this function will not be used within PAL as it handles curves separately?

File: src/core/textrenderer/qgstextrenderer.h [hunk 2]
```diff
@@ -264,6 +299,15 @@ class CORE_EXPORT QgsTextRenderer
      */
     static constexpr double FONT_WORKAROUND_SCALE = 10;
 
+    // to match QTextEngine handling of superscript/subscript font sizes
```
Comment (nirvn) [inline src/core/textrenderer/qgstextrenderer.h:302]: Is that comment a copy/paste artifact?

File: src/core/textrenderer/qgstextrenderer.h [hunk 3]
```diff
@@ -144,6 +144,41 @@ class CORE_EXPORT QgsTextRenderer
                           QgsRenderContext &context, const QgsTextFormat &format,
                           bool drawAsOutlines = true );
 
+    /**
+     * Draws text along a line using the specified settings.
+     *
+     * \param line line to render text along
```
Comment (nirvn) [inline src/core/textrenderer/qgstextrenderer.h:150]: The line points are defined in painter unit, correct? Should we specify in the doc or self-obvious? Same Q for the next function.

File: src/core/textrenderer/qgstextrenderer.h (hunk 2 above)
Comment (nyalldawson) [reply src/core/textrenderer/qgstextrenderer.h:302]: No, it relates to the constant value below -- SUPERSCRIPT_SUBSCRIPT_FONT_SIZE_SCALING_FACTOR

File: src/core/textrenderer/qgstextrenderer.h (hunk 3 above)
Comment (nyalldawson) [reply src/core/textrenderer/qgstextrenderer.h:150]: I don't think it's harmful to be overly verbose even for "obvious" things! If you're unsure, then a mere mortal will definitely be unsure!

**Comment by nyalldawson** [comment]
@nirvn all addressed!

**Review by nirvn — APPROVED** [review]

**Pull request merged**